### ISSUE TITLE
What about F# support?

### ISSUE THREAD

**Issue by ShalokShalom** [FlaxEngine/FlaxEngine#587]
Since C# is supported, is there also F# support?

It compiles to the same CIL and runs on Mono and the new .Net.

**Comment by TheWitheredStriker** [comment]
Quick status update.

I've successfully made Flax's C# API available to F#. I did so by following much of the same steps in the link, but for Flax, i. e.:

1. Create an F# class library in the Flax project from the commandline in its own folder (here FSharp) and build it into a DLL.
2. Move the FSharp.Core.dll file over to `<ProjectName>\Source/Game`.
3. After opening the `.fsproj` file in Visual Studio, right-click Dependencies -> Add Project Reference. Browse to `C:\Program Files (x86)\Flax\Flax_1.2\Binaries\Editor\Win64\Development` and add a reference to `FlaxEngine.CSharp.dll`. Apply.
4. Replace the current code in Library.fs with that from the screenshot below, build (Ctrl - Shift - B) and copy the output FSharp.dll from `...\Flax Projects\<ProjectName>\FSharp\bin\Debug\net5.0` to `<ProjectName>\Source/Game`.

![image](https://user-images.githubusercontent.com/38614982/138109012-81121b1f-1e15-40d7-893a-0685574d0a96.png)

An additional few screenshots to prove the availability of the API to F#:

![image](https://user-images.githubusercontent.com/38614982/138109157-f581d614-fba2-48b0-8eec-32bc5c225ef8.png)
![image](https://user-images.githubusercontent.com/38614982/138109356-bc78b818-b949-4dbb-973b-c60a3eab4196.png)

I also opened up the C# API to Visual Basic .NET for good measure, showing that it too is theoretically compatible:

![image](https://user-images.githubusercontent.com/38614982/138113576-970853ff-984c-49e4-a4a6-94ad4b2aadd7.png)

While I find it pretty awesome that all of this works, I haven't actually got the script to execute in Flax. Unlike in Unity, where I could open the DLL in the content browser to find the script and then drag it onto something, Flax will not let me do any of that. This could be because I erred somewhere, but it could also be the result of Flax ... well, being Flax, and not Unity. They're not the same engine, after all, and work quite differently.

That being said, I still firmy believe that it's possible to use F#, VB.NET and all other CLR languages (Like ClojureCLR, Boo and Ada for .NET, to name a few) with Flax. We'll just need to figure out the correct approach, which will naturally be a more painful journey than with Unity given Flax's young age, its currently small (but growing!) community, and the fact that we're basically the first people to look into it, like, ever.

I do believe that it'll be worth investigating further, even if just for the sake of science and fun. The fact that I could access the Flax API in F# and Visual Basic to begin with is already an interesting discovery and a good step forward in my opinion. If anyone else makes any further advancements in this field, like maybe getting Flax to actually run our code, don't hestitate to let us know. I'll continue to report any additional findings in the meantime.

**Comment by TheWitheredStriker** [comment]
Hello again, Happy New Year to all of you, hoping you are well :)

https://christianfindlay.com/2020/10/17/how-to-use-fsharp-and-csharp/ Would something like this work? 

![image](https://user-images.githubusercontent.com/38614982/148600172-49d17f20-5e5f-4eb8-b875-def7f06fe937.png)

I tried to build this into a DLL, reference it in Flax's C# code, and then create a new instance of `FlaxFS.Hi()`, but I can't get it to work.

```
C:\Users\vande\Documents\Flax Projects\CLRTest\Source\Game\Main.cs(18,17,18,23): error CS0246: The type or namespace name 'FlaxFS' could not be found (are you missing a using directive or an assembly reference?
```

I DID reference the assembly: 

![image](https://user-images.githubusercontent.com/38614982/148600633-433c16e1-6919-42d5-92f7-69fc9b63f151.png)

Regardless, I may be onto a dirty, yet working way of scripting Flax games in other .NET languages (having the bulk of the code in F# / VB.NET / etc files and then calling those functions and classes from C#). I haven't identified the problem yet, but I might be onto something.

Would this be a feasible workaround?

**Comment by mafiesto4** [comment]
Yeah building F# into a separate dll and just using it as a reference to the main game assembly could be a nice workaround. You can add this reference inside `Game.Build.cs` script setup method:
```cs
options.ScriptingAPI.FileReferences.Add(Path.Combine("path/to/assembly.dll"));
```

**Comment by TheWitheredStriker** [comment]
Awesome! Thank you! I'll try that out and post my results here, as usual.

**Comment by mafiesto4** [comment]
We won't be working on this (as official support). Not enough requests and not enough man-power.
